### PR TITLE
Bug 1274130 - Validate data when processing firebase records

### PR DESCRIPTION
**Pull request by mancas**

### CODE REVIEW

File: shared/js/dataDriver.js [hunk 1]
```diff
@@ -437,5 +462,25 @@ loop.DataDriver = function() {
         request.send(JSON.stringify(body) || null);
       });
     }
+
+    _validateData(type, data) {
+      let schema = this._getSchemaByType(type);
+      let validatedData = new loop.validate.Validator(schema).validate(data || {});
```
Comment (Mardak) [inline shared/js/dataDriver.js]: We should always have data, so if it's not there, we should probably fail.

File: shared/js/dataDriver.js [hunk 2]
```diff
@@ -437,5 +462,25 @@ loop.DataDriver = function() {
...
+
+      return validatedData;
```
Comment (Mardak) [inline shared/js/dataDriver.js]: Did you mean to do something with this return value?

File: shared/js/dataDriver.js [hunk 3]
```diff
@@ -437,5 +462,25 @@ loop.DataDriver = function() {
...
+    _getSchemaByType(type) {
+      switch (type) {
+        case "chat":
+          return CHAT_VALIDATOR_SCHEMA;
+        case "participant":
+          return PARTICIPANT_VALIDATOR_SCHEMA;
+        case "presence":
+          return PRESENCE_VALIDATOR_SCHEMA;
+        default:
+          return {};
```
Comment (Mardak) [inline shared/js/dataDriver.js]: An unexpected type should probably make the record invalid.

File: shared/js/dataDriver.js [hunk 4]
```diff
@@ -7,6 +7,22 @@ loop.DataDriver = function() {
...
+  /*
+   * When creating new schemas, be sure to add them to _getSchemaByType method
+   */
+  const CHAT_VALIDATOR_SCHEMA = {
+    timestamp: Number,
+    value: Object
+  };
+
+  const PARTICIPANT_VALIDATOR_SCHEMA = {
+    value: Object
```
Comment (Mardak) [inline shared/js/dataDriver.js]: Any reason no timestamp here? All schemas have to be one level deep for Validator, and all records should have at most two fields of timestamp and optionally value (for deleted records).

File: shared/test/dataDriver_test.js [hunk 5]
```diff
@@ -455,9 +459,35 @@ describe("loop.DataDriver", () => {
...
+    });
+
+    it("should throw an exception if value isn't valid", () => {
+      driver._processRecord("chat!01234567abcdefghijkl", {
+        timestamp: 1234567890123,
+        value: 12412
+      });
+
+      sinon.assert.calledOnce(console.error);
+    });
```
Comment (Mardak) [inline shared/test/dataDriver_test.js:527]: Add a test for `timestamp: NaN` where `typeof NaN` is "number".

File: shared/js/dataDriver.js [hunk 6]
```diff
@@ -360,10 +376,19 @@ loop.DataDriver = function() {
      *  - {Mixed}  value     Stored value for the record.
      */
     _processRecord(key, data) {
-      let dispatchAction;
-      let dispatchExtra = {};
       // XXX akita bug 1274130: Validate data/key/timestamp before processing.
       let [, type, id] = key.match(/^([^!]+)!(.+)$/);
```
Comment (Mardak) [inline shared/js/dataDriver.js]: The `key` should be validated although you could probably just do this by saving the result of `match` to check if it's `null` before doing the destructured assignment.

File: shared/js/dataDriver.js [hunk 7]
```diff
@@ -360,10 +376,19 @@ loop.DataDriver = function() {
      *  - {Mixed}  value     Stored value for the record.
      */
     _processRecord(key, data) {
-      let dispatchAction;
-      let dispatchExtra = {};
       // XXX akita bug 1274130: Validate data/key/timestamp before processing.
```
Comment (Mardak) [inline shared/js/dataDriver.js]: Don't forget to delete this XXX comment when we're done with this bug! :)

File: shared/js/dataDriver.js (hunk 6 above)
Comment (Mardak) [inline shared/js/dataDriver.js]: Alternatively declare `let type, id;` outside of the try and just do the destructured assignment on the `match` where if there's no match, there'll be a `TypeError: key.match(...) is null`

File: shared/js/dataDriver.js [hunk 8]
```diff
@@ -360,10 +376,19 @@ loop.DataDriver = function() {
...
       let [, type, id] = key.match(/^([^!]+)!(.+)$/);
+      try {
+        // First validates data, if no exception is thrown we can dispatch the actions
+        this._validateData(type, data);
+      } catch (exception) {
+        console.error("Validate data error: ", exception);
```
Comment (Mardak) [inline shared/js/dataDriver.js]: We're technically validating not just "data" but the "record" which has the key (including type) and the data (including timestamp and value). ;)

File: shared/test/dataDriver_test.js [hunk 9]
```diff
@@ -473,9 +503,22 @@ describe("loop.DataDriver", () => {
           userId: "theUserId"
         }));
     });
+
+    it("should throw an exception if value isn't valid", () => {
+      driver._processRecord("participant!theUserId", {
+        timestamp: 1234567890123,
+        value: 1234
```
Comment (Mardak) [inline shared/test/dataDriver_test.js:567]: Technically the value can be of any value. E.g., if we had a `meta!roomName` record, the value would probably be a string and not an object.

Edit: Oh, but I see this is specifically for the participant.

**Comment by Standard8** [comment]
Support for Hello/Loop has been discontinued.

https://support.mozilla.org/kb/hello-status

Hence closing the old bugs. Thank you for your support.

**Pull request closed without merging**